### PR TITLE
feat: validate the passing object whether it's a `RegExp`

### PR DESCRIPTION
**Pull request by hyf0**
## Summary

- I also moved the `js_values` and `ext` modules to `rspack_napi_shared` crate for better cohesion.

When you passing wrong value

image

### <samp>🤖 Generated by Copilot at f0c8df6</samp>

Refactored the `rspack_napi_shared` crate to provide common functionality and types for the `rspack_binding_options` and `rspack_binding_webpack` crates. Moved some files and modules from the `rspack_binding_options` crate to the `rspack_napi_shared` crate and updated the imports accordingly. Added a type check for JavaScript RegExp objects and a utility function for calling `Object.prototype.toString`.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at f0c8df6</samp>

*  Move `ext` and `js_values` modules from `rspack_binding_options` to `rspack_napi_shared` crate ([link](https://github.com/web-infra-dev/rspack/pull/3495/files?diff=unified&w=0#diff-75b5c77f5846e4148f21e2b14150e14f587a7d0db6a6642730958c934271a021L3-L4), [link](https://github.com/web-infra-dev/rspack/pull/3495/files?diff=unified&w=0#diff-2c00d4f66acdce3babd851a1b2091cafc0c3e4377f545c6eae7a98555495b077), [link](https://github.com/web-infra-dev/rspack/pull/3495/files?diff=unified&w=0#diff-a84d6a5b28ba76d482c3736df820ccabbd3095a2cbb2eaa099e68ab5c0372876), [link](https://github.com/web-infra-dev/rspack/pull/3495/files?diff=unified&w=0#diff-55cb0c7efeb4995b617d0d320b33213b0a7c9f558fa4df41a56dae54d17691e6))
*  Import `JsRegExp` and `JsStringExt` types from `rspack_napi_shared` crate in `raw_split_chunks.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3495/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968R7))
*  Remove unused imports of `JsRegExp` and `JsStringExt` from `crate` scope in `raw_split_chunks.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3495/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L149-L150))
*  Add `#![feature(try_blocks)]` attribute and `ext`, `js_values`, and `utils` modules to `rspack_napi_shared` crate ([link](https://github.com/web-infra-dev/rspack/pull/3495/files?diff=unified&w=0#diff-bf94c064fa0e67d217aea942b4d1886ff0cafd9d2d48473d7b3ecd0126fc412cL1-R6))
*  Re-export `JsStringExt`, `JsRegExp`, and `object_prototype_to_string_call` items from `rspack_napi_shared` crate ([link](https://github.com/web-infra-dev/rspack/pull/3495/files?diff=unified&w=0#diff-bf94c064fa0e67d217aea942b4d1886ff0cafd9d2d48473d7b3ecd0126fc412cR15-R19))
*  Add `object_prototype_to_string_call` function to `utils.rs` in `rspack_napi_shared` crate ([link](https://github.com/web-infra-dev/rspack/pull/3495/files?diff=unified&w=0#diff-7b1d6cec1d9036706764f6dda48337688e3234737f33aa2832c56c5b0084e8e4R1-R29))
*  Add assertion to check type of JavaScript RegExp object in `js_reg_exp.rs` in `rspack_napi_shared` crate ([link](https://github.com/web-infra-dev/rspack/pull/3495/files?diff=unified&w=0#diff-ba486e4e8fea226d18222485ad4ef1c5e1431d9274389de6ce0314485fcc552cR8-R9), [link](https://github.com/web-infra-dev/rspack/pull/3495/files?diff=unified&w=0#diff-ba486e4e8fea226d18222485ad4ef1c5e1431d9274389de6ce0314485fcc552cL37-R40))

</details>

### CODE REVIEW

**Review by ahabhgk — APPROVED** [review]

File: crates/rspack_napi_shared/src/js_values/js_reg_exp.rs [hunk 1]
```diff
@@ -34,7 +36,8 @@ impl FromNapiValue for JsRegExp {
     napi_val: napi::sys::napi_value,
   ) -> napi::Result<Self> {
     let js_object = JsObject::from_raw(env, napi_val)?;
-    // TODO(hyf0): we should do dome validation here. Such as make sure the type of the Object is `[object RegExp]`
+    let ty_string = object_prototype_to_string_call(env, &js_object)?;
+    assert_eq!(ty_string.as_str(), "[object RegExp]");
```
Comment (hyf0) [inline crates/rspack_napi_shared/src/js_values/js_reg_exp.rs:40]: I should return an `Err` here. This breaks the try mechanism of `Either`.

Comment (hyf0) [reply crates/rspack_napi_shared/src/js_values/js_reg_exp.rs]: Done.

**Comment by ahabhgk** [comment]
I'm ok with adding validate logic on js -> rust, but we should validate everything by schema, and report all invalidate options before it passes to rust side

**Comment by hyf0** [comment]
> I'm ok with adding validate logic on js -> rust, but we should validate everything by schema, and report all invalidate options before it passes to rust side or before normalization

Schema validation could be skipped. 

However, it's more of an enhancement of `napi` rather than schema validation. We need a way to ensure the correctness and semantics of customized napi values defined by Rspack.

**Review by ahabhgk — APPROVED** [review]

**Pull request merged**